### PR TITLE
Superseded: Bug 626576 AppSource workaround

### PR DESCRIPTION
**Pull request by martinsrui-msft**
## Superseded

Investigation showed that this must be fixed in the Platform extension-operation model rather than through BCApps error handling.

The BCApps workaround and its test have been removed, leaving this branch with no net application changes. The root fix is now tracked by:

https://microsoft.ghe.com/bic/BC-Platform/pull/46282

[AB#626576](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/626576)

### CODE REVIEW

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al [hunk 1]
```diff
@@ -272,6 +276,7 @@ codeunit 2501 "Extension Marketplace"
             end else begin
                 ExtensionPendingSetup.SetRange("User Id", UserSecurityId());
                 ExtensionPendingSetup.DeleteAll();
+                ShowAppInstallationError(InstallErrorText);
```
[the remark this answers is not in the document]
Comment (martinsrui-msft) [reply src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al]: Not changing this path: AL `Message()` is deferred until AL execution finishes; it does not pause at this line like `Confirm` or `StrMenu`. Microsoft documents that the message is displayed only after the AL code finishes (https://learn.microsoft.com/dynamics365/business-central/dev-itpro/developer/methods-auto/dialog/dialog-message-method). Therefore `DeleteAll()` does not hold locks across user dismissal, and adding `Commit()` here would introduce an unnecessary transaction boundary.

(Learning for future reviews: apply the prompt-inside-transaction rule to immediate interactions such as `Confirm`/`StrMenu`, not deferred `Message()` calls.)

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al [hunk 2]
```diff
@@ -366,10 +376,18 @@ codeunit 2501 "Extension Marketplace"
...
     begin
-        HasSucceeded := ExtensionInstallationImpl.InstallExtensionWithConfirmDialog(PackageId, lcid);
+        ClearLastError();
+        if not TryInstallApp(PackageId, lcid, HasSucceeded) then begin
+            InstallErrorText := GetLastErrorText();
+            MakeMarketplaceTelemetryCallback(ResponseURL, OperationResult::DeploymentFailedDueToPackage);
+            if InstallErrorText = '' then
+                InstallErrorText := AppInstallationFailedMsg;
+            // Preserve the outer TryFunction failure result so its caller can clean up before showing the error.
+            Error(InstallErrorText);
```
[the remark this answers is not in the document]
Comment (martinsrui-msft) [reply src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al]: Fixed in `dcb60a4b21`: the rethrow now uses a stable `Label` as the first `Error()` argument and passes the original platform detail as a substitution value, preserving useful UI detail without putting customer-derived text in the RT0030 message field.

File: src/System Application/Test/Extension Management/src/ExtensionMarketplaceTest.Codeunit.al [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    end;
+
+    [ModalPageHandler]
+    procedure InstallMarketplaceExtensionPageHandler(var MarketplaceExtnDeployment: TestPage "Marketplace Extn Deployment")
+    begin
+        MarketplaceExtnDeployment.Continue.Invoke();
+        MarketplaceExtnDeployment.Install.Invoke();
+    end;
+
+    [MessageHandler]
```
[the remark this answers is not in the document]
Comment (martinsrui-msft) [reply src/System Application/Test/Extension Management/src/ExtensionMarketplaceTest.Codeunit.al]: Fixed in `3c8e0e93af` and retained in `dcb60a4b21`: each test enqueues the expected message, the handler dequeues it through `Assert.ExpectedMessage`, and `LibraryVariableStorage.AssertEmpty()` verifies that it fired exactly once.

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinsrui-msft) [reply src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al]: Not changing this path: AL `Message()` is deferred until AL execution finishes; it does not pause at this line like `Confirm` or `StrMenu`. Microsoft documents that the message is displayed only after the AL code finishes (https://learn.microsoft.com/dynamics365/business-central/dev-itpro/developer/methods-auto/dialog/dialog-message-method). Therefore `DeleteAll()` does not hold locks across user dismissal, and adding `Commit()` here would introduce an unnecessary transaction boundary.

(Learning for future reviews: apply the prompt-inside-transaction rule to immediate interactions such as `Confirm`/`StrMenu`, not deferred `Message()` calls.)

File: src/System Application/Test/Extension Management/src/ExtensionMarketplaceTest.Codeunit.al [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        PermissionsMock: Codeunit "Permissions Mock";
+        FailingAppId: Guid;
+        DetailedInstallFailureMsg: Label 'The test extension installation failed with a detailed error.';
+        MicrosoftPublisherTxt: Label 'Microsoft', Locked = true;
+        FixtureNotPublishedErr: Label 'The extension test fixture %1 must be published and tenant-visible before running this test.', Comment = '%1 = app ID';
+        FixtureInstalledErr: Label 'The extension test fixture must not be installed before running this test.';
+        ExtensionInstalledErr: Label 'The extension should not be installed.';
+        PendingSetupNotClearedErr: Label 'The failed installation should clear pending extension setup for the current user.';
+
+    [Test]
```
[the remark this answers is not in the document]
Comment (martinsrui-msft) [reply src/System Application/Test/Extension Management/src/ExtensionMarketplaceTest.Codeunit.al]: Partially addressed in `dcb60a4b21`: `FailedFirstPartyInstallByMarketplaceIdShowsOriginalError` now covers the text marketplace-ID overload using the same real installer fixture. I am not adding a blank-error test because the platform installer supplies an error for real failures; forcing an empty value would require a second artificial fixture or a production-only test seam. The defensive fallback remains a single guarded branch, while the real error and cleanup paths have integration coverage.

**Comment by martinsrui-msft** [comment]
Addressed the latest automated review in `d5c53ba6ee` by reducing the PR to the active Guid path and one focused test:

- **Performance / `Message()` after `DeleteAll()`** - No code change. AL defers `Message()` until execution completes, so it does not hold the write transaction across user dismissal like `Confirm` or `StrMenu` would. Adding `Commit()` would introduce an unnecessary transaction boundary.
- **XML documentation** - Obsolete after this commit: all newly added public test-library helpers were removed.
- **Modal-page handler verification** - Obsolete after this commit: the fixture integration flow and modal handler were removed.
- **Wrapped expected-message mismatch** - Obsolete after this commit: the inner try/rethrow and wrapped error label were removed. The remaining test verifies the error captured directly by the existing outer TryFunction.

The final net diff now contains only `ExtensionMarketplace.Codeunit.al` and `ExtensionMarketplaceTest.Codeunit.al`.

**Comment by thloke** [comment]
This is the wrong fix, it's at best just a workaround. Our extension model has all extension installations recorded under the Extension Management page. The error should be recorded there, just the same as the error we get today when trying to install an extension that is not supported in the current locale for example:

image
